### ISSUE TITLE
Tokenizer is thread-unsafe from v0.6.9

### ISSUE THREAD

**Issue by KosukeArase** [WorksApplications/sudachi.rs#289]
Hi, thank you for developing and maintaining this great project!

We recently noticed that tokenizer is no longer thread-safe on v0.6.9.
Here is the minimal reproducible example, it works 0.6 <= sudachipy <= 0.6.8 but causes `RuntimeError: Already borrowed` on v0.6.9.

```python
import concurrent.futures
from concurrent.futures import ThreadPoolExecutor

from sudachipy import dictionary


TOKENIZER = dictionary.Dictionary().create()


def analyze(txt: str):
    return TOKENIZER.tokenize(txt)


def test(n: int = 1000, txt: str = "高輪ゲートウェイ駅"):
    with ThreadPoolExecutor(max_workers=2) as executor:
        futures = [executor.submit(analyze, txt) for _ in range(n)]

        for future in concurrent.futures.as_completed(futures):
            print(future.result())

test()
```

You can reproduce it by running [this example](https://colab.research.google.com/drive/1CC-GB7Cgb-kSmNyJDI0kePTsDE7kFmmc).

**Comment by eiennohito** [comment]
Tokenizers were never thread safe. 0.6.9 has changed rust<-> python interop
story a bit, so the lack of thread safety is more clearly exposed now.
There were locks before on tokenizer usage. Please use one Tokenizer per
thread either by putting them into a thread local variable or pooling them.

The only exception to thread safety story is pre_tokenizer() which is
thread safe and supports multithreaded usage by tokenizers.

To summarize:
Dictionary -> thread safe, immutable after construction
Tokenizer -> not thread safe, tokenization changes mutable state
MorphermeList -> not thread safe wrt to concurrent tokenization when used
as out parameter
PosFilter -> thread safe, immutable
pre_tokenizer() -> thread safe, supports parallel usage by tokenizers
library internally

**Comment by KosukeArase** [comment]
That makes sense to me, thank you for the clear explanation!
